agent: data race in endpoint.(*Endpoint).writeL4Map()

```
Dec 27 09:34:15 anam cilium-agent[23721]: ==================
Dec 27 09:34:15 anam cilium-agent[23721]: WARNING: DATA RACE
Dec 27 09:34:15 anam cilium-agent[23721]: Read at 0x00c420ed3650 by goroutine 42:
Dec 27 09:34:15 anam cilium-agent[23721]:   runtime.mapiterinit()
Dec 27 09:34:15 anam cilium-agent[23721]:       /home/scanf/opt/go/src/runtime/hashmap.go:709 +0x0
Dec 27 09:34:15 anam cilium-agent[23721]:   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).writeL4Map()
Dec 27 09:34:15 anam cilium-agent[23721]:       /home/scanf/opt/og/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:55 +0xab
Dec 27 09:34:15 anam cilium-agent[23721]:   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).writeL4Policy()
Dec 27 09:34:15 anam cilium-agent[23721]:       /home/scanf/opt/og/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:113 +0x2b5
Dec 27 09:34:15 anam cilium-agent[23721]:   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).writeHeaderfile()
Dec 27 09:34:15 anam cilium-agent[23721]:       /home/scanf/opt/og/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:256 +0x1fa8
Dec 27 09:34:15 anam cilium-agent[23721]:   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).regenerateBPF()
Dec 27 09:34:15 anam cilium-agent[23721]:       /home/scanf/opt/og/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:433 +0x218
Dec 27 09:34:15 anam cilium-agent[23721]:   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).regenerate()
Dec 27 09:34:15 anam cilium-agent[23721]:       /home/scanf/opt/og/src/github.com/cilium/cilium/pkg/endpoint/policy.go:693 +0x4ca
Dec 27 09:34:15 anam cilium-agent[23721]:   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).Regenerate.func1()
Dec 27 09:34:15 anam cilium-agent[23721]:       /home/scanf/opt/og/src/github.com/cilium/cilium/pkg/endpoint/policy.go:759 +0x2a1
Dec 27 09:34:15 anam cilium-agent[23721]: Previous write at 0x00c420ed3650 by goroutine 109:
Dec 27 09:34:15 anam cilium-agent[23721]:   runtime.mapassign_faststr()
Dec 27 09:34:15 anam cilium-agent[23721]:       /home/scanf/opt/go/src/runtime/hashmap_fast.go:598 +0x0
Dec 27 09:34:15 anam cilium-agent[23721]:   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).writeL4Map()
Dec 27 09:34:15 anam cilium-agent[23721]:       /home/scanf/opt/og/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:71 +0x5e5
Dec 27 09:34:15 anam cilium-agent[23721]:   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).writeL4Policy()
Dec 27 09:34:15 anam cilium-agent[23721]:       /home/scanf/opt/og/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:113 +0x2b5
Dec 27 09:34:15 anam cilium-agent[23721]:   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).writeHeaderfile()
Dec 27 09:34:15 anam cilium-agent[23721]:       /home/scanf/opt/og/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:256 +0x1fa8
Dec 27 09:34:15 anam cilium-agent[23721]:   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).regenerateBPF()
Dec 27 09:34:15 anam cilium-agent[23721]:       /home/scanf/opt/og/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:433 +0x218
Dec 27 09:34:15 anam cilium-agent[23721]:   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).regenerate()
Dec 27 09:34:15 anam cilium-agent[23721]:       /home/scanf/opt/og/src/github.com/cilium/cilium/pkg/endpoint/policy.go:693 +0x4ca
Dec 27 09:34:15 anam cilium-agent[23721]:   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).Regenerate.func1()
Dec 27 09:34:15 anam cilium-agent[23721]:       /home/scanf/opt/og/src/github.com/cilium/cilium/pkg/endpoint/policy.go:759 +0x2a1
Dec 27 09:34:15 anam cilium-agent[23721]: Goroutine 42 (running) created at:
Dec 27 09:34:15 anam cilium-agent[23721]:   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).Regenerate()
Dec 27 09:34:15 anam cilium-agent[23721]:       /home/scanf/opt/og/src/github.com/cilium/cilium/pkg/endpoint/policy.go:742 +0x1e6
Dec 27 09:34:15 anam cilium-agent[23721]:   github.com/cilium/cilium/pkg/endpointmanager.TriggerPolicyUpdates.func1()
Dec 27 09:34:15 anam cilium-agent[23721]:       /home/scanf/opt/og/src/github.com/cilium/cilium/pkg/endpointmanager/manager.go:270 +0x289
Dec 27 09:34:15 anam cilium-agent[23721]: Goroutine 109 (running) created at:
Dec 27 09:34:15 anam cilium-agent[23721]:   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).Regenerate()
Dec 27 09:34:15 anam cilium-agent[23721]:       /home/scanf/opt/og/src/github.com/cilium/cilium/pkg/endpoint/policy.go:742 +0x1e6
Dec 27 09:34:15 anam cilium-agent[23721]:   github.com/cilium/cilium/pkg/endpointmanager.TriggerPolicyUpdates.func1()
Dec 27 09:34:15 anam cilium-agent[23721]:       /home/scanf/opt/og/src/github.com/cilium/cilium/pkg/endpointmanager/manager.go:270 +0x289
Dec 27 09:34:15 anam cilium-agent[23721]: ==================
Dec 27 09:34:15 anam cilium-agent[23721]: WARNING: DATA RACE
Dec 27 09:34:15 anam cilium-agent[23721]: Read at 0x00c4202ee488 by goroutine 42:
Dec 27 09:34:15 anam cilium-agent[23721]:   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).writeL4Map()
Dec 27 09:34:15 anam cilium-agent[23721]:       /home/scanf/opt/og/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:55 +0x124
Dec 27 09:34:15 anam cilium-agent[23721]:   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).writeL4Policy()
Dec 27 09:34:15 anam cilium-agent[23721]:       /home/scanf/opt/og/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:113 +0x2b5
Dec 27 09:34:15 anam cilium-agent[23721]:   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).writeHeaderfile()
Dec 27 09:34:15 anam cilium-agent[23721]:       /home/scanf/opt/og/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:256 +0x1fa8
Dec 27 09:34:15 anam cilium-agent[23721]:   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).regenerateBPF()
Dec 27 09:34:15 anam cilium-agent[23721]:       /home/scanf/opt/og/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:433 +0x218
Dec 27 09:34:15 anam cilium-agent[23721]:   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).regenerate()
Dec 27 09:34:15 anam cilium-agent[23721]:       /home/scanf/opt/og/src/github.com/cilium/cilium/pkg/endpoint/policy.go:693 +0x4ca
Dec 27 09:34:15 anam cilium-agent[23721]:   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).Regenerate.func1()
Dec 27 09:34:15 anam cilium-agent[23721]:       /home/scanf/opt/og/src/github.com/cilium/cilium/pkg/endpoint/policy.go:759 +0x2a1
Dec 27 09:34:15 anam cilium-agent[23721]: Previous write at 0x00c4202ee488 by goroutine 109:
Dec 27 09:34:15 anam cilium-agent[23721]:   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).writeL4Map()
Dec 27 09:34:15 anam cilium-agent[23721]:       /home/scanf/opt/og/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:71 +0x607
Dec 27 09:34:15 anam cilium-agent[23721]:   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).writeL4Policy()
Dec 27 09:34:15 anam cilium-agent[23721]:       /home/scanf/opt/og/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:113 +0x2b5
Dec 27 09:34:15 anam cilium-agent[23721]:   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).writeHeaderfile()
Dec 27 09:34:15 anam cilium-agent[23721]:       /home/scanf/opt/og/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:256 +0x1fa8
Dec 27 09:34:15 anam cilium-agent[23721]:   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).regenerateBPF()
Dec 27 09:34:15 anam cilium-agent[23721]:       /home/scanf/opt/og/src/github.com/cilium/cilium/pkg/endpoint/bpf.go:433 +0x218
Dec 27 09:34:15 anam cilium-agent[23721]:   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).regenerate()
Dec 27 09:34:15 anam cilium-agent[23721]:       /home/scanf/opt/og/src/github.com/cilium/cilium/pkg/endpoint/policy.go:693 +0x4ca
Dec 27 09:34:15 anam cilium-agent[23721]:   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).Regenerate.func1()
Dec 27 09:34:15 anam cilium-agent[23721]:       /home/scanf/opt/og/src/github.com/cilium/cilium/pkg/endpoint/policy.go:759 +0x2a1
Dec 27 09:34:15 anam cilium-agent[23721]: Goroutine 42 (running) created at:
Dec 27 09:34:15 anam cilium-agent[23721]:   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).Regenerate()
Dec 27 09:34:15 anam cilium-agent[23721]:       /home/scanf/opt/og/src/github.com/cilium/cilium/pkg/endpoint/policy.go:742 +0x1e6
Dec 27 09:34:15 anam cilium-agent[23721]:   github.com/cilium/cilium/pkg/endpointmanager.TriggerPolicyUpdates.func1()
Dec 27 09:34:15 anam cilium-agent[23721]:       /home/scanf/opt/og/src/github.com/cilium/cilium/pkg/endpointmanager/manager.go:270 +0x289
Dec 27 09:34:15 anam cilium-agent[23721]: Goroutine 109 (running) created at:
Dec 27 09:34:15 anam cilium-agent[23721]:   github.com/cilium/cilium/pkg/endpoint.(*Endpoint).Regenerate()
Dec 27 09:34:15 anam cilium-agent[23721]:       /home/scanf/opt/og/src/github.com/cilium/cilium/pkg/endpoint/policy.go:742 +0x1e6
Dec 27 09:34:15 anam cilium-agent[23721]:   github.com/cilium/cilium/pkg/endpointmanager.TriggerPolicyUpdates.func1()
Dec 27 09:34:15 anam cilium-agent[23721]:       /home/scanf/opt/og/src/github.com/cilium/cilium/pkg/endpointmanager/manager.go:270 +0x289
Dec 27 09:34:15 anam cilium-agent[23721]: ==================
```

```
$ git ll
b7090303c272 (Ginkgo: redirects vagrant provision output to GinkgoWriter, 2017-12-26)
$ git diff
diff --git a/Makefile.defs b/Makefile.defs
index eb50e28d49b9..3ba7307c122f 100644
--- a/Makefile.defs
+++ b/Makefile.defs
@@ -11,10 +11,10 @@ BUILD = $(VERSION) $(shell git show -s --format='format:%h %aD') $(shell go vers
 GOBUILD = -ldflags '-X "github.com/cilium/cilium/pkg/version.Version=$(BUILD)"'
 
 # Uncomment to enable race detection
-#GOBUILD += -race
+GOBUILD += -race
 
 # Uncomment to enable deadlock detection
-#GOBUILD += -tags lockdebug
+GOBUILD += -tags lockdebug
 
 ifneq ($(LOCKDEBUG),)
     GOBUILD += -tags lockdebug
[...]
```

If I enable race detection, my build fails..

```
go build -i -ldflags '-X "github.com/cilium/cilium/pkg/version.Version=1.0.0-rc4 7721f1e74e30 2018-01-19T10:52:58-08:00 go version go1.9 linux/amd64"' -race -o cilium-agent
# github.com/cilium/cilium/daemon
/usr/local/go/pkg/tool/linux_amd64/link: running gcc failed: exit status 1
collect2: error: ld returned 1 exit status

Makefile:13: recipe for target 'cilium-agent' failed
```

--- a/pkg/endpoint/bpf.go
+++ b/pkg/endpoint/bpf.go
@@ -52,7 +52,7 @@ func (e *Endpoint) writeL4Map(fw *bufio.Writer, owner Owner, m policy.L4PolicyMa
 	array := ""
 	index := 0
 
-	for k, l4 := range m {
+	for _, l4 := range m {
 		// Represents struct l4_allow in bpf/lib/l4.h
 		protoNum, err := u8proto.ParseProtocol(string(l4.Protocol))
 		if err != nil {
@@ -62,15 +62,6 @@ func (e *Endpoint) writeL4Map(fw *bufio.Writer, owner Owner, m policy.L4PolicyMa
 		dport := byteorder.HostToNetwork(uint16(l4.Port))
 
 		redirect := uint16(l4.L7RedirectPort)
-		if l4.IsRedirect() && redirect == 0 {
-			redirect, err = e.addRedirect(owner, &l4)
-			if err != nil {
-				return err
-			}
-			l4.L7RedirectPort = int(redirect)
-			m[k] = l4
-		}
-
 		redirect = byteorder.HostToNetwork(redirect).(uint16)
 		entry := fmt.Sprintf("%d,%d,%d,%d", index, dport, redirect, protoNum)
 		if array != "" {
@@ -405,6 +396,25 @@ func (e *Endpoint) updateCT(owner Owner, flushEndpointCT bool, consumersAdd, con
 	return wg
 }
 
+// allocateRedirects must be called while holding the endpoint and consumable
+// locks for writing. On success, returns nil; otherwise, returns an error
+// indicating the problem that occurred while adding an l7 redirect for the
+// specified policy.
+func (e *Endpoint) allocateRedirects(owner Owner, m policy.L4PolicyMap) error {
+	for k, l4 := range m {
+		redirect := uint16(l4.L7RedirectPort)
+		if l4.IsRedirect() && redirect == 0 {
+			redirect, err := e.addRedirect(owner, &l4)
+			if err != nil {
+				return err
+			}
+			l4.L7RedirectPort = int(redirect)
+			m[k] = l4
+		}
+	}
+	return nil
+}
+
 // regenerateBPF rewrites all headers and updates all BPF maps to reflect the
 // specified endpoint.
 // Must be called with endpoint.Mutex not held and endpoint.BuildMutex held.
@@ -428,6 +438,26 @@ func (e *Endpoint) regenerateBPF(owner Owner, epdir, reason string) (uint64, err
 		e.getLogger().WithField(logfields.EndpointState, e.state).Debug("Skipping build due to invalid state")
 		e.Mutex.Unlock()
 		return 0, fmt.Errorf("Skipping build due to invalid state: %s", e.state)
+	}
+
+	// If there is a Consumable, walk the L4Policy for ports that require
+	// an L7 redirect and add them to the endpoint; update the L4PolicyMap
+	// with the redirects.
+	if e.Consumable != nil {
+		e.Consumable.Mutex.Lock()
+		if e.Consumable.L4Policy != nil {
+			if err = e.allocateRedirects(owner, e.Consumable.L4Policy.Ingress); err != nil {
+				e.Consumable.Mutex.Unlock()
+				e.Mutex.Unlock()
+				return 0, fmt.Errorf("Unable to allocate ingress redirects: %s", err)
+			}
+			if err = e.allocateRedirects(owner, e.Consumable.L4Policy.Egress); err != nil {
+				e.Consumable.Mutex.Unlock()
+				e.Mutex.Unlock()
+				return 0, fmt.Errorf("Unable to allocate egress redirects: %s", err)
+			}
+		}
+		e.Consumable.Mutex.Unlock()
 	}
 
 	if err = e.writeHeaderfile(epdir, owner); err != nil {

